File: cpp/LeetCode/No102.cpp
/*
Given the root of a binary tree, return the level order traversal of its nodes' values. 
(i.e., from left to right, level by level).
*/
#include <iostream>
#include <vector>
#include <queue>
#include "Debug\TreeNode.h"
#include "Debug\VectorDebug.h"

using namespace std;

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if(!root) return res;
        queue<TreeNode*> q;
        q.push(root);
        while(q.size()>0){
            vector<int> temp;
            int len = q.size();
            for(int i=0;i<len;i++){
                TreeNode* node = q.front();
                q.pop();
                temp.push_back(node->val);
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
            res.push_back(temp);
        }
        return res;
    }
};

int main(){
    Solution solution;

    TreeNode* root = new TreeNode(3, new TreeNode(9), new TreeNode(20, new TreeNode(15), new TreeNode(7)));
    print2DVector<int>(solution.levelOrder(root));
    return 0;
}

File: python/LeetCode Quiz/No102.py
from typing import List
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        q = [root]
        level = 0
        res = []
        while q:
            size = len(q)
            for _ in range(size):
                node = q.pop(0)
                if len(res)==level:
                    res.append([node.val])
                else:
                    res[level].append(node.val)
                
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            level+=1
        return res

solution = Solution()
root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
print(solution.levelOrder(root))